### PR TITLE
Specify core.p4 match_kind semantics

### LINKED ISSUE #1090: The core.p4 match kinds are not described in the P416 spec (but they are in PSA!)
This is a relatively minor issue with the P416 specification, but one that I found confusing.

The match kinds in core.p4 (exact, ternary, lpm) are not really explained in the P416 spec, apart from some very terse comments in Appendix D:
https://p4.org/p4-spec/docs/P4-16-working-spec.html#sec-p4-core-lib
No big deal; it's not hard to infer what these mean. 

But, I was surprised to find the complete explanations of the core.p4 match kinds in the latest version of the PSA spec, here:
https://p4.org/p4-spec/docs/PSA.html#sec-match-kinds

Please consider moving the descriptions of exact, ternary, and lpm match from the PSA spec to Appendix D of the P416 spec.

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>
Fixes #1090 
To be discussed at the LDWG

### CODE REVIEW

**Comment by apinski-cavium** [comment]
+1 from me

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6592]: Would it be reasonable to generalize this to "all types for which equality is defined" ?

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6599]: "then" -> "when" ?

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6609]: typoe: "entires" -> "entries"

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:6592]: Does the control plane support such types for keys? 
Perhaps the control-plane is the scope of a separate document, though, and this spec should be very liberal.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6602]: a mask with only zeros is valid

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6600]: "higher" is implementation-dependent

**Comment by jafingerhut** [comment]
@jafingerhut AI for Andy: check whether anywhere in the spec already says that when a table has multiple keys, an apply() operation only matches an entry if ALL the individual field matching criteria are true.  That is, logically there is an AND between the matching conditions between multiple keys of a table.

**Comment by mbudiu-vmw** [comment]
Can we get some more comments on this PR? We would like to merge this in the next design meeting.
I will apply the suggested changes.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6640]: "this choice is left to the target implementation".  It seems more precise to say that "this choice is determined by the run time API used"

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6828]: I am not fundamentally against adding this phrase "as described in Section [#sec-table-keys}" here, but I do not see anywhere in that section that specifies how one key set expression may turn into one or multiple entries in a table.  It seems reasonable to me that the language specification NOT try to specify that level of detail.

Basically, the addition of this phrase seems confusing to me.  If I go where it says to go, it does not explain what the sentence before it claims it will.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6648]: Missing period at end of sentence.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6649]: I do not know if this is the best place in the spec to say this, but it seems worth saying at least once somewhere in the spec something like the following:

"The text above defines the condition when one key lookup field matches one field within an entry added to the table.  If a table has multiple key field defined, then a lookup key only matches an entry if the match condition for all fields is true simultaneously for a single table entry."

I am betting there is probably a clearer way to state that.

File: p4-16/spec/P4-16-spec.mdk
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk:6632]: Is this a sufficient description of mask? Or should we expand?

**Comment by jafingerhut** [comment]
@jafingerhut @vgurevich Pinging us to remind us to review the latest version of this PR, last updated on 2022-Nov-04.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6650]: this needs square braces

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6651]: Comment on "and a priority".  One might get the impression from reading this that if the table has 3 ternary fields, then there are 3 priorities.  There are not.  There is only one priority value for the entire table entry.

One way to avoid causing such confusion would be not to mention priorities except in relation to an entire table entry, later.  When describing the match condition for a single field, omit any mention of priority.

Thus this bullet item could be replaced with:
```
- a `ternary` match kind on a key field means that the field in the
  table specifies a set of values for the key field using a value and
  a mask.  The meaning of the `(value, mask)` pair is
  similar to the P4 mask expressions, as described in Section [#sec-cubes]:
  a key field `k` matches the table entry when `k & mask == value & mask`.
```

Then later after all of the bullet points that describe exact, ternary, and lpm, only THEN bring up the issue that some table entries require a priority (those with at least one `ternary` field), and some table entries do not require a priority (those with no `ternary` fields).  The priority is not used to determine which entries match, only which of multiple matching entries "wins", and its action is executed.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6674]: There is a sentence in the current PR addressing this comment, so I am resolving this.

**Comment by mbudiu-vmw** [comment]
@jafingerhut we are supposed to merge this but we need an approval from you

**Comment by jafingerhut** [comment]
I did add a couple of review comments on 2022-Dec-05, and at least one of them I consider serious enough NOT to approve this PR in its current form (the one about the implied priority-per-field, which is untrue).

I would prefer if changes could be made to this PR to resolve those comments before I approve, but if others want to approve this as it is, I won't stop them.

**Comment by mbudiu-vmw** [comment]
Sorry, lost track which comments have been addressed.
Any others I should be heeding?

**Comment by mbudiu-vmw** [comment]
@jafingerhut hopefully this is better. I had some conflicts when rebasing, hopefully I didn't break anything.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6314]: `exact` -> `ternary`

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6503]: My previous comment still seems not addressed.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6325]: Still missing period at end of sentence.

**Comment by mbudiu-vmw** [comment]
@jafingerhut maybe this is closer to a resolution

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: The cross-reference has been removed, addressing my comment.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: Updated text in commit 2 resolves this comment.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: confirmed that this has been fixed.

**Review by jafingerhut — APPROVED** [review]
LGTM.  I believe all of my (numerous) comments have been resolved.

**Pull request merged**
